src/bases/createWidgetBase.ts

```diff
@@ -0,0 +1,201 @@
+
+interface WidgetInternalState {
+	readonly id?: string;
+	dirty: boolean;
+	widgetClasses: string[];
+	cachedWidget?: VNode;
+	historicChildrenMap: Map<string | Factory<Widget<WidgetState>, WidgetOptions<WidgetState>>, Widget<WidgetState>>;
+	currentChildrenMap: Map<string | Factory<Widget<WidgetState>, WidgetOptions<WidgetState>>, Widget<WidgetState>>;
+};
+
+/**
+ * Internal state map for widget instances
+ */
+const widgetInternalStateMap = new WeakMap<Widget<WidgetState>, WidgetInternalState>();
+
+/**
+ * The counter for generating a unique ID
+ */
+let widgetCount = 0;
+
+function generateID(instance: Widget<WidgetState>): string {
+	const id = `widget-${++widgetCount}`;
+	instance.setState({ id });
+	return id;
+}
+
+function isWNode(child: DNode): child is WNode {
+	return child && (<WNode> child).factory !== undefined;
+}
+
+function realizeDNode(instance: Widget<WidgetState>, dNode: DNode) {
+	const internalState = widgetInternalStateMap.get(instance);
+	let child: HNode | Widget<WidgetState>;
+	if (isWNode(dNode)) {
+		const { factory, options: { id, state } } = dNode;
+		const childrenMapKey = id || factory;
+		const cachedChild = internalState.historicChildrenMap.get(childrenMapKey);
+		if (cachedChild) {
+			child = cachedChild;
+			if (state) {
+				child.setState(state);
+			}
+		} else {
+			child = factory(dNode.options);
+			child.own(child.on('invalidate', () => {
+				instance.invalidate();
+			}));
+			internalState.historicChildrenMap.set(childrenMapKey, child);
+			instance.own(child);
+		}
+		if (!id && internalState.currentChildrenMap.has(factory)) {
+			console.error('must provide unique keys when using the same widget factory multiple times');
+		}
+		internalState.currentChildrenMap.set(childrenMapKey, child);
+	}
+	else {
+		child = dNode;
+		if (child.children) {
+			child.children = child.children.map((child: DNode) => {
+				return realizeDNode(instance, child);
+			});
+		}
+	}
+	return child.render();
+}
+
+function manageDetachedChildren(instance: Widget<WidgetState>) {
+	const internalState = widgetInternalStateMap.get(instance);
+
+	internalState.historicChildrenMap.forEach((child, key) => {
+		if (!internalState.currentChildrenMap.has(key)) {
+			internalState.historicChildrenMap.delete(key);
+			child.destroy();
+		}
+	});
+	internalState.currentChildrenMap.clear();
+}
+
+const createWidget: WidgetFactory = createStateful
+	.mixin<WidgetMixin, WidgetOptions<WidgetState>>({
+		mixin: {
+			childNodeRenderers: [],
+
+			classes: [],
+
+			getChildrenNodes(this: Widget<WidgetState>): DNode[] {
+				let childrenWrappers: DNode[] = [];
+
+				this.childNodeRenderers.forEach((fn) => {
+					const wrappers = fn.call(this);
+					childrenWrappers = childrenWrappers.concat(wrappers);
+				});
+
+				return childrenWrappers;
+			},
+
+			getNodeAttributes(this: Widget<WidgetState>, overrides?: VNodeProperties): VNodeProperties {
+				const props: VNodeProperties = {};
+
+				this.nodeAttributes.forEach((fn) => {
```

Comment:
It just returns the attributes from the mixin and then mixes them over the existing props.